internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/TagContext.java

```diff
@@ -175,6 +175,15 @@ public void putTag(final String key, final String value) {
     this.tags.set(key, value);
   }
 
+  public void putTagIfAbsent(final String key, final String value) {
+    if (this.tags == null) {
+      this.tags = TagMap.create(4);
+    }
+    if (!this.tags.containsKey(key)) {
+      this.tags.set(key, value);
+    }
+  }
```

Comment:
I would keep the `tags` inflation in only place by reusing the existing `putTag()` method:

```suggestion
    if (this.tags == null || !this.tags.containsKey(key)) {
      putTag(key, value);
    }
  }
```